File: nodejs/packages/periph/src/connection/hx711_mock.js
'use strict';

/**
 * In-memory fake HX711 connection for unit tests — no hardware, no GPIO.
 *
 * HX711 has no register map; each conversion is instead selected by the
 * pulse count (25/26/27) sent to `readRaw`. Preload signed 24-bit conversion
 * results with `queueRead`; each `readRaw` call pops the next one (0 if the
 * queue is empty) and, matching every real connection's contract, validates
 * `numPulses` is one of {25, 26, 27}, throwing an `Error` otherwise. Every
 * call's pulse count is appended to `reads` so tests can assert which
 * channel/gain a driver call actually requested (e.g. that `setGain` drives
 * the right pulse count). `powerDown`/`powerUp` calls are logged to
 * `powerCalls`.
 */
class HX711ConnectionMock {
    constructor() {
        this._queue = [];
        this.reads = [];
        this.powerCalls = [];
        this.ready = true;
    }

    /** Queue a signed 24-bit value to be returned by the next readRaw() call. */
    queueRead(value) {
        this._queue.push(value);
    }

    isReady() {
        return this.ready;
    }

    readRaw(numPulses) {
        if (numPulses !== 25 && numPulses !== 26 && numPulses !== 27)
            throw new Error('numPulses must be 25, 26, or 27');
        this.reads.push(numPulses);
        if (this._queue.length > 0) return this._queue.shift();
        return 0;
    }

    powerDown() {
        this.powerCalls.push('down');
    }

    powerUp() {
        this.powerCalls.push('up');
    }

    close() {}
}

module.exports = { HX711ConnectionMock };

File: python/periph/connection/hx711_mock.py
class HX711ConnectionMock:
    """In-memory fake HX711 connection for unit tests — no hardware, no GPIO.

    HX711 has no register map; each conversion is instead selected by the
    pulse count (25/26/27) sent to ``read_raw``. Preload signed 24-bit
    conversion results with ``queue_read``; each ``read_raw`` call pops the
    next one (0 if the queue is empty) and, matching every real connection's
    contract, validates ``num_pulses`` is one of {25, 26, 27}, raising
    ``ValueError`` otherwise. Every call's pulse count is appended to
    ``reads`` so tests can assert which channel/gain a driver call actually
    requested (e.g. that ``set_gain`` drives the right pulse count).
    ``power_down``/``power_up`` calls are logged to ``power_calls``.
    """

    def __init__(self):
        self._queue = []
        self.reads = []
        self.power_calls = []
        self.ready = True

    def queue_read(self, value):
        """Queue a signed 24-bit value to be returned by the next read_raw()."""
        self._queue.append(value)

    def is_ready(self):
        return self.ready

    def read_raw(self, num_pulses):
        if num_pulses not in (25, 26, 27):
            raise ValueError("num_pulses must be 25, 26, or 27")
        self.reads.append(num_pulses)
        if self._queue:
            return self._queue.pop(0)
        return 0

    def power_down(self):
        self.power_calls.append('down')

    def power_up(self):
        self.power_calls.append('up')

    def close(self):
        pass
